Remove (dfs_)query_and_fetch from the REST API

aleph-zero opened the pull request:
Remove the ability to specify search type ‘query_and_fetch’ and ‘dfs_query_and_fetch’ from the REST API.

- Adds REST tests
- Updates REST API spec to remove ‘query_and_fetch’ and ‘dfs_query_and_fetch’ as options
- Removes documentation for these options

Closes #9606

jpountz commented on src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java:
```diff
@@ -96,8 +97,18 @@ public static SearchRequest parseSearchRequest(RestRequest request) {
+        String searchType = request.param("search_type");
+        if (SearchType.fromString(searchType).equals(SearchType.QUERY_AND_FETCH) ||
+                SearchType.fromString(searchType).equals(SearchType.DFS_QUERY_AND_FETCH)) {
+            throw new ElasticsearchIllegalArgumentException("Unsupported search type [" + searchType + "]");
```
Can you make SearchType.fromString called only once?

aleph-zero replied:
What would that achieve?

jpountz replied:
I'm a bit nitpicking but if we can avoid calling this parsing method twice easily I think we should do it? Additionally I tend to find that extracting intermediate results to local variables helps readability.

jpountz replied:
By the way I was just looking at call sites of `SearchType.fromString` and it looks to me that removing `(dfs_)query_and_fetch` from there would have the effect that you expect without having to add this special case to RestSearchAction? This way it would also have the good side effect of applying the same restriction to the `more_like_this` action.

aleph-zero replied:
Agree that we could simply remove ```(dfs_)query_and_fetch``` from ```SearchType.fromString```. However, I feel that this would be the wrong approach. ```SearchType.fromString``` is a pure utility function that performs a single, well-defined task: Given a string it returns the corresponding enum type. Modifying the function to reflect the needs of an external, context-dependent use case strikes me as dangerous. It bakes an assumption that may not always hold into an otherwise simple, easy-to-understand function.

With respect to catching other cases such as ```more_like_this```, perhaps we might consider adding some function to ```SearchType``` to check if it is allowed to be called from public API's, e.g. ```SearchType.isInternalUseOnly```, or some such. Of course that opens the new problem that now users of the class must know about this distinction and check for it where appropriate. 

Thoughts?

jpountz replied:
> SearchType.fromString is a pure utility function that performs a single, well-defined task:

Maybe we can change its name/contract?

> we might consider adding some function to SearchType to check if it is allowed to be called from public API's, e.g. SearchType.isInternalUseOnly

My preference would go to having two enums: a public one for consumers of our APIs and an internal one to let elasticsearch know how it should execute the query. But I think we should defer it to another issue.

jpountz commented:
+1 on making them purely internal

Should we do something about the Java client API too?

aleph-zero commented:
@jpountz I was wondering about the Java API, but the ticket is pretty explicit on it being a REST issue. I suppose we can leave it in Java for 'power' users?

jpountz commented:
I asked the question because I think these options are equally trappy for consumers of the REST API and of the Java API. It might be harder to fix since the SearchType class is something we both need internally (so can't really remove these QUERY_AND_FETCH types ) but I think we should at least fix the documentation to make them sound less cool and label as internal options explicitely?

aleph-zero commented:
I removed the remaining references to these two options from the reference part of the docs. I searched the docs for the Java API, but did not find any reference to SearchType.(DFS_)QUERY_AND_FETCH, so I added a 'NOTE' about them being purely for internal purposes.

jpountz commented on docs/reference/search/request-body.asciidoc:
```diff
@@ -64,8 +64,8 @@ And here is a sample response:
-    `dfs_query_then_fetch`, `dfs_query_and_fetch`, `query_then_fetch`,
-    `query_and_fetch`. Defaults to `query_then_fetch`. See
+    `dfs_query_then_fetch`, `query_then_fetch`, 'count',
+    or 'scan'. Defaults to `query_then_fetch`. See
```
`count` should not be added back as it is deprecated

s1monw commented:
+1

jpountz commented:
LGTM

clintongormley commented:
@aleph-zero you want to get this in?
